### ISSUE TITLE
Mark invalid in 1.12.2, not in 1.12.1

### ISSUE THREAD

**Issue by yuhlzu** [jhy/jsoup#1325]
Document docTime = Jsoup.parse(s)

```bash
 W/System.err: org.jsoup.UncheckedIOException: java.io.IOException: Mark invalid
 W/System.err:     at org.jsoup.parser.CharacterReader.rewindToMark(CharacterReader.java:148)
 W/System.err:     at org.jsoup.parser.Tokeniser.consumeCharacterReference(Tokeniser.java:192)
 W/System.err:     at org.jsoup.parser.TokeniserState$38.read(TokeniserState.java:759)
 W/System.err:     at org.jsoup.parser.Tokeniser.read(Tokeniser.java:59)
 W/System.err:     at org.jsoup.parser.TreeBuilder.runParser(TreeBuilder.java:55)
 W/System.err:     at org.jsoup.parser.TreeBuilder.parse(TreeBuilder.java:47)
 W/System.err:     at org.jsoup.parser.Parser.parse(Parser.java:107)
 W/System.err:     at org.jsoup.Jsoup.parse(Jsoup.java:58)
```

**Comment by jhy** [comment]
Can you check on HEAD - fixed an issue relating to this in #1324

Or, post code so that we can replicate.

**Comment by rturner-edjuster** [comment]
I am seeing the same or a similar issue (where 1.12.1 works for my automated tests, but 1.12.2 does not). In my case, HEAD (50fae5bf) resolves the problem for my test cases.

@jhy Is there a chance of making a 1.12.3 version that just has this patch (unless of course you anticipate 1.13.1 coming fairly soon)?

If neither a 1.12.3 or 1.13.1 is possible or imminent, would you recommend going back to 1.12.1 or back to 1.11.3? (we don't need anything specific from 1.12.x at this point, but from reading comments, I believe there are also issues with 1.11.3 related to this fix/issue)

**Comment by rturner-edjuster** [comment]
Although after running with 1.12.1 on our build server, I have that it doesn't work either for our tests. I'll fall back to 1.11.3 for now.

**Comment by jhy** [comment]
Can you get me the files / code that is failing tests? And passing in HEAD. It will help build confidence that it's corrected. Are the cases you see failing different between .1 and 2? If they are the same -- then something else is at play I think.

**Comment by rturner-edjuster** [comment]
I'm pretty sure they were very different failures. I had 3 failures on my dev machine for 1.12.2, but quite a lot for 1.12.1 on our build server (Linux), but none on my development machine (OSX). I will try to pull together some of the input data for you -- might take me a few hours at least (and I have to just fix another issue for something unrelated before I can get to it tonight).

**Comment by jhy** [comment]
@rturner-edjuster to your question on a release - I'm thinking that I'll cherry-pick this specific fix and release it as 1.12.3. For 1.13 I want to make some larger changes (have removed some deprecated classes, looking to refactor for multi-project support, etc). Would like to get something out in the next few days, so people don't bump into this without a forward path.

**Comment by jhy** [comment]
@yuhlzu still would be good to hear from you too if possible

**Comment by yuhlzu** [comment]
the html file from okhttps will case error, but the html file from my computer is ok
please use this code quickly,  the cookie will be overdue 
```java
 String url = "http://jwk.lzu.edu.cn/academic/teacher/teachingtask/student_list.jsdo?epid=308941312";

        Request request = new Request
                .Builder()
                .url(url)
                .header("Cookie", "JSESSIONID=3DD0DFDA9237B502E7C3E343AA3B2ADA.TA2")
                .build();
// 同步请求
        try {

            Response response = new OkHttpClient().newCall(request).execute();
            if (response.code() == 200) {
                ResponseBody rb = response.body();
                if (rb != null) {
                    String s = rb.string();

                    System.out.println(s);
//                        LogUtils.i(s);

                    Document docTime = Jsoup.parse(s);
                    
//                        LogUtils.i("=================" + docTime.toString());
                    Elements elements = docTime.getElementsByClass("infolist_hr_common");

                    System.out.println(elements.size());
                } else {
                    System.out.println(response.message() + response.code());
                }
            }
        } catch (IOException e) {
            System.out.println(e.toString());
            e.printStackTrace();
        }
```

**Comment by yuhlzu** [comment]
> @yuhlzu still would be good to hear from you too if possible

**Comment by yuhlzu** [comment]
> @yuhlzu still would be good to hear from you too if possible

**Comment by rturner-edjuster** [comment]
@jhy Email sent to your published address on Github -- I have some data ready for you. Please see the email for details.

**Comment by jhy** [comment]
@yuhlzu I just get timeouts trying to fetch that. Jsoup isn't involved in that request path so you could just save the file directly and attach it here. Could you do that please?

And also, can you confirm if you are getting the issue on HEAD?

**Comment by rturner-edjuster** [comment]
Also, in case @yuhlzu might find this useful, here's some code I used to "dump" the failed data from my tests:
`    private static String readBundleVersionFromManifestForClass(Class clazz) {
        try (InputStream is = clazz.getClassLoader().getResourceAsStream(JarFile.MANIFEST_NAME)) {
            Manifest m = new Manifest(is);
            return StringUtils.defaultString(m.getMainAttributes().getValue("Bundle-Version"), "unknown-version");
        } catch (IOException ex) {
            return "exception-version";
        }
    }

    private static Document jsoupParse(String input) {
        try {
            return Jsoup.parse(input);
        } catch (org.jsoup.UncheckedIOException ex) {
            if (ex.getMessage().contains("Mark invalid")) {
                try {
                    // Write the failing input to a file
                    final String jsoupVersionString = readBundleVersionFromManifestForClass(Jsoup.class);
                    final String dateTimeString = DateTimeFormatter.ofPattern("yyyyMMdd_HHmmssSSS").withZone(ZoneId.systemDefault()).withLocale(Locale.CANADA).format(Instant.now());
                    final String filename = "jsoup_" + jsoupVersionString + "_mark_invalid_" + dateTimeString + ".html";
                    FileUtils.writeStringToFile(new File(filename), input, StandardCharsets.UTF_8);
                    LOG.error("Jsoup mark invalid data written to: {}", filename);
                } catch (IOException e) {
                    LOG.error("Failed to write Jsoup failed parsing input data to file", e);
                }
            }
            throw ex; // Make sure we fail as we would have
        }
    }
`

**Comment by yuhlzu** [comment]
> @yuhlzu我只是尝试获取超时而已。Jsoup不在该请求路径中，因此您可以直接保存文件并将其附加在此处。你能这样做吗？
> 
> 另外，您可以确认是否在HEAD上遇到问题吗？

you got timeouts, maybe I in China, or you can set more time to request or by vpn

I find this error: now cookie is 
```
JSESSIONID=0B48F448C26BB2D51FD547D01A81A238.TA2
```

Caused by: java.io.IOException: Mark invalid

Jousp is ok   in that request path, like 
```
http://jwk.lzu.edu.cn/academic/teacher/teachingtask/student_list.jsdo?epid=308939361
```

but is error like
```
http://jwk.lzu.edu.cn/academic/teacher/teachingtask/student_list.jsdo?epid=308941312
```

**Comment by jhy** [comment]
OK, I confirmed with the extra files that @rturner-edjuster sent offline that the issue seen was the same as what we fixed in #1324. I added an extra test to validate that if an invalid character reference is found at the buffer boundary and the buffer is marked, and therefore rewound, that the parser successfully reads the invalid character reference, emits it as such, and safely carries on.

19be649e596b3066ed0d98d009cdd5b73d14c08d

Will add some extra cases around when parsed from a String, not a File.

I'll leave this issue open, although it's fixed, so that others will find it more readily vs re-opening.
